From 50a206223f061e3628f9c8e2279dff04049e3d07 Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Wed, 11 Dec 2013 04:02:33 +0000
Subject: Trivial typo fix

---
(limited to 'gtk/gtkheaderbar.c')

diff --git a/gtk/gtkheaderbar.c b/gtk/gtkheaderbar.c
index 2e5f3e2..3fba3c2 100644
--- a/gtk/gtkheaderbar.c
+++ b/gtk/gtkheaderbar.c
@@ -1729,7 +1729,7 @@ gtk_header_bar_class_init (GtkHeaderBarClass *class)
   g_object_class_install_property (object_class,
                                    PROP_SUBTITLE,
                                    g_param_spec_string ("subtitle",
-                                                        P_("Subitle"),
+                                                        P_("Subtitle"),
                                                         P_("The subtitle to display"),
                                                         NULL,
                                                         G_PARAM_READWRITE));
--
cgit v0.9.2

